src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt

```diff
@@ -284,9 +310,15 @@ internal class ModelCheckingStrategy(
         }
 
         fun chooseThread(iThread: Int): Int =
-            if (nextThreadToSwitch.hasNext()) {
+            if (currentInterleavingPosition < threadSwitchChoices.size) {
+                check(executionPosition == -1 || (executionPosition == 0 && loopDetector.replayModeEnabled) || executionPosition == switchPositions[currentInterleavingPosition - 1]) {
```

Comment:
Perhaps makes sense to handle case `executionPosition == -1` separately. 
In this case the loop detector controls the scheduling, so what should `chooseThread` return?
I think returning just `threadSwitchChoices[currentInterleavingPosition++]` is incorrect?